File: LeetCode/C++/3. Longest Substring Without Repeating Characters.cpp
///LeetCode 3. Longest Substring Without Repeating Characters by C++
///write by LazyWolfLin in 16.09.10

/*
class Solution
{
public:
    bool letter[128];
    int lengthOfLongestSubstring(string s)
    {
        size_t l=0,r=0;
        int ans=0;
        memset(letter,false,sizeof(letter));
        while (l<s.size())
        {
            while (r<s.size())
            {
                if (letter[s[r]]) break;
                letter[s[r]]=true;
                if (r-l+1>ans) ans=r-l+1;
                r++;
            }
            while (l!=r)
            {
                letter[s[l]]=false;
                l++;
                if (s[l-1]==s[r]) break;
            }
        }
        return ans;
    }
};
*/

class Solution
{
public:
    int letter[128];
    int lengthOfLongestSubstring(string s)
    {
        int l=0,r=0;
        int ans=0;
        memset(letter,0,sizeof(letter));
        while (r<s.size())
        {
            l=max(l,letter[s[r]]);
            ans=max(ans,r-l+1);
            letter[s[r]]=r+1;
            r++;
        }
        return ans;
    }
};

File: LeetCode/Python3/3. Longest Substring Without Repeating Characters.py
# LeetCode 3. Longest Substring Without Repeating Characters by Python3
# write by LazyWolfLin in 18.07.23

class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = -1
        ans = 0
        letter = dict()
        for x, ch in enumerate(s):
            if ch in letter and letter[ch] > start:
                start = letter[ch]
            ans = max(ans, x - start)
            letter[ch] = x
        return ans